### ISSUE TITLE
Warn when casting the address of an enum variant constructor to integer

### ISSUE THREAD

**Issue by Boddlnagg** [rust-lang-nursery/rust-clippy#2588]
Assume the following code:
```rust
#[derive(Debug)]
enum Foo {
    A,
    B
}

fn main() {
   let x = Foo::A;
   let y = x as i32;
   
   println!("{:?}", y);
}
```

This prints `0`.

I then change `Foo::A` to take some additional payload (as part of a refactoring, etc), such as
```rust
#[derive(Debug)]
enum Foo {
    A(usize),
    B
}
```

Now the cast in `main` still compiles without warnings, but suddenly returns the address of the enum constructor function `Foo::A`.

It might even be reasonable to have this as a warning in `rustc` itself (but I'm not sure how to formulate the conditions to prevent false positives), but if not, at least clippy could warn.

**Comment by oli-obk** [comment]
We should lint all function pointer/function name casts to integers other than `usize`

**Comment by maximiliangraebel** [comment]
I would like to take a look at this one as my first issue. Any tips on where to start?

**Comment by oli-obk** [comment]
You can add a new check after https://github.com/rust-lang-nursery/rust-clippy/blob/8ec61a613ad1278b18a4770332b6da128704fdd6/clippy_lints/src/types.rs#L902 for function pointer to integer casts. There's a lot of code there for checking types, so you can probably find examples for figuring out whether some types is `usize` or `isize`. For the function pointer type you'll need to match on the type and check for `TyFnDef` or `TyFnPtr`
